design/dynamic-provision-filesystem.md

```diff
@@ -0,0 +1,196 @@
+
+## Current Experience
+
+To provision and consume a filesystem, the experience is the following:
+
+To create a Filesystem, the following Filesystem CRD is created
+
+```yaml
+apiVersion: rook.io/v1alpha1
+kind: Filesystem
+metadata:
+  name: myfs
+  namespace: rook
+spec:
+  metadataPool:
+    replicated:
+      size: 3
+  dataPools:
+    - erasureCoded:
+       codingChunks: 2
+       dataChunks: 2
+  metadataServer:
+    activeCount: 1
+    activeStandby: true
+```
+
+Users will have to come up with these values and ensure every parameter is provided correctly.
+
+To consume it, the pod manifest has to have the following:
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: mysql
+spec:
+  strategy:
+    type: Recreate
+  template:
+    spec:
+      containers:
+      - image: mysql:5.6
+        name: mysql
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          value: changeme
+        ports:
+        - containerPort: 3306
+          name: mysql
+        volumeMounts:
+        - name: mysql-persistent-storage
+          mountPath: /var/lib/mysql
+      volumes:
+      - name: mysql-persistent-storage
+        cephfs:
+          monitors:
+          - monitor1
+          - monitor2
+          - monitor3
+          user: admin
+          secretRef:
+            name: rook-admin
+```
+
+## Experience with Dynamic Provisioned Filesystem
+
+To create a filesystem, you just create a PVC object. This is consistent with all other storage provisioning in Kubernetes.
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: myfs
+spec:
+  storageClassName: rook-filesystem-simple
+  accessModes:
+  - ReadWriteMany
+```
+
+To consume it, the pod manifest is shown as follows:
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: mysql
+spec:
+  strategy:
+    type: Recreate
+  template:
+    spec:
+      containers:
+      - image: mysql:5.6
+        name: mysql
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          value: changeme
+        ports:
+        - containerPort: 3306
+          name: mysql
+        volumeMounts:
+        - name: mysql-persistent-storage
+          mountPath: /var/lib/mysql
+      volumes:
+      - name: mysql-persistent-storage
+        persistentVolumeClaim:
+          claimName: myfs
+```
+
+Note that the consuming pod manifest looks the same whether it is mounting a filesystem or a block device.
+
+## StorageClass Example
+
+Notice that there was a reference to a StorageClass called `rook-filesystem-simple` in the filesystem PVC example I previously showed. Dynamic provisioned storage refers to a StorageClass object for details and configuration about how the storage should be provisioned.
+The storage class is setup by the administrator and can look as follows for filesystem:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-filesystem-simple
+provisioner: rook.io/filesystem
+parameters:
+  metadataPool:
```

Comment:
I would prefer for the admin to still create a separate `Filesystem` CRD and then set the name in the parameters, and not have the "whole" `Filesystem` CRD Spec in the parameters.

Having it separate allows us to change the CRD independently of the `StorageClass`, so the `StorageClass` will "always" stay the same (only containing the FS name as a parameter).

Reply:
So you are advocating to admins creating filesystems only and users just consume them. As supposed to users creating the filesystems

Reply:
Yes. I prefer that users are not creating `Filesystem`s (even through a `StorageClass`).
I as the admin want to create the `Filesystem`s and `StorageClass`es for the `Filesystem`s.
The users then consume the admin created `StorageClass`es.

Reply:
ok.. i think you got a good point. I will change the design proposal to behave as follows:
- Admin create filesystem CRD and StorageClass. Filesystem name is referenced in StorageClass
- Users consume filesystems by creating PVCs. Users can also provide a `path` for mounting a specify directory in cephfs
@galexrt does that work better?

Reply:
@kokhang Yes, that is better. Thank you! :+1: